### PR TITLE
ui: bazel build for UI assets (db-console and cluster-ui)

### PR DESCRIPTION
**Pull request by koorosh**
- build `cluster-ui` JS package
- build `db-console` package
- build `crdb-protobufjs-client` package
- embed JS assets into `distccl` go's binary

Depends on PR: https://github.com/cockroachdb/cockroach/pull/69135

Release justification: non-production code changes

### CODE REVIEW

File: pkg/ui/BUILD.bazel
Comment (koorosh) [inline pkg/ui/BUILD.bazel:17]: get all `node_module`'s packages except local linked packages

Comment (rickystewart) [reply pkg/ui/BUILD.bazel:17]: `node_modules` is a `gitignore`d directory, so this seems wrong. What are you trying to do here?

**Review by rickystewart — CHANGES_REQUESTED** [review]

File: pkg/ui/BUILD.bazel
Comment (koorosh) [reply pkg/ui/BUILD.bazel:17]: Bazel treats `node_modules` as an external repository and used to include external dependencies:
```
workspace(
    name = "cockroach",
    managed_directories = {
       "@npm": ["pkg/ui/node_modules"]
    },
)
```
except external packages, it also included symlinked local packages
```
|- workspaces
    |- cluster-ui
    |- db-console
|- node_modules
    |- ... 
    |- cluster-ui (symlink) 
```
and to avoid circular dependencies for cluster-ui rules, we need all packages from node_modules except `cluster-ui`.

**Comment by rickystewart** [comment]
bazel side looks fine so far, i'll leave the remainder of the review to someone who's knowledgeable about UI stuff.

**Comment by koorosh** [comment]
@rickystewart , can you help me to resolve following issue with CI build? It fails due to error with writing files to filesystem:
> error Could not write file "/go/src/github.com/cockroachdb/cockroach/pkg/ui/yarn-error.log": "EROFS: read-only file system, open '/go/src/github.com/cockroachdb/cockroach/pkg/ui/yarn-error.log'"
  error An unexpected error occurred: "EROFS: read-only file system, mkdir '/go/src/github.com/cockroachdb/cockroach/pkg/ui/node_modules'".

I've managed to resolve the same problem locally by adding `rw` permissions to mounted docker volume (for `dev builder`) (https://github.com/cockroachdb/cockroach/pull/67066/commits/14d34ab9a4f783f83c5a89df3ed14f6f8defe5a6) but it didn't help on Teamcity.

This issue happens because bazel installs `node_modules` in host environment and then links it inside of sandboxed env.

**Comment by rickystewart** [comment]
@koorosh Good catch, I'll send out a PR to fix that in a moment.

**Comment by rickystewart** [comment]
https://github.com/cockroachdb/cockroach/pull/69643

File: WORKSPACE
Comment (rickystewart) [inline WORKSPACE:6]: Do the other `node_modules` directories (`pkg/ui/workspaces/cluster-ui/node_modules`, `pkg/ui/workspaces/db-console/node_modules`) need to be in this list as well?

File: pkg/ui/distccl/BUILD.bazel
Comment (rickystewart) [inline pkg/ui/distccl/BUILD.bazel:7]: Nit: `-R` is more standard than `-r` here.

File: pkg/ui/distccl/BUILD.bazel
Comment (rickystewart) [inline pkg/ui/distccl/BUILD.bazel:4]: Bad style for a rule to have the same name as an output file, so you should give the rule a different name (e.g. `genassets`).

File: pkg/ui/distccl/BUILD.bazel
Comment (rickystewart) [inline pkg/ui/distccl/BUILD.bazel:6]: So generally you [should](https://docs.bazel.build/versions/main/be/general.html#general-advice) avoid creating directories as `genrule` outputs:

> Avoid creating symlinks and directories. Bazel doesn't copy over the directory/symlink structure created by genrules and its dependency checking of directories is unsound.

And indeed when I run the build I get this warning trying to build this target:

    WARNING: /cockroach/pkg/ui/distccl/BUILD.bazel:3:8: output 'pkg/ui/distccl/assets' of //pkg/ui/distccl:genassets is a directory; dependency checking of directories is unsound

For *now*, if we have to do this, this is fine and we can check it in so we can unblock other work, but we would want to fix this before we can call this production-ready. A couple options:

1. You can declare all the sub-files under the `assets` directory explicitly. I see in that directory there are a lot of files with long unpredictable names (like `02290aae2a70e2d8dc8d391049661b25.woff2`), so maybe this isn't possible? 
2. You can compress the `assets` directory into a single file (like a `.tar.gz`), and then in the Go code you can decompress the `.tar.gz` and expose those artifacts as an `FS`. [Very small example.](https://stackoverflow.com/questions/51073133/bazel-genrule-that-outputs-a-directory)

(2) is probably what you'll have to do, although it'll take a little refactoring on the Go side. Like I said, we can take that as a TODO for later.

File: pkg/ui/distoss/BUILD.bazel
Comment (rickystewart) [inline pkg/ui/distoss/BUILD.bazel:4]: Same comments as above: should have distinct `name`, directory output is unsound, use `cp -R` instead of `cp -r`

File: pkg/ui/workspaces/db-console/src/app.tsx
Comment (rickystewart) [inline pkg/ui/workspaces/db-console/src/app.tsx:19]: Does the UI build (as built with `make`) still work even with this commented out?

File: pkg/ui/workspaces/cluster-ui/webpack.config.js
Comment (rickystewart) [inline pkg/ui/workspaces/cluster-ui/webpack.config.js:149]: Same question -- can this safely be commented out without breaking the `make` build?

**Review by rickystewart — CHANGES_REQUESTED** [review]
I didn't look too closely at the JS changes, you'll want to get another reviewer to make sure those changes are sound.

File: pkg/ui/distccl/BUILD.bazel
Comment (rickystewart) [reply pkg/ui/distccl/BUILD.bazel:6]: Oh, there might be a low-friction third option that allows us to accomplish the same goal without a bunch of refactoring. Not really sure, I'd have to look closer. Happy to have a call about this if you want to.

**Comment by rickystewart** [comment]
Oh, rebase atop `master` to get rid of those conflicts/fix the Docker mount issue.

File: WORKSPACE
Comment (koorosh) [reply WORKSPACE]: Done, thanks!

File: pkg/ui/distccl/BUILD.bazel
Comment (koorosh) [reply pkg/ui/distccl/BUILD.bazel]: Done

File: pkg/ui/distccl/BUILD.bazel
Comment (koorosh) [reply pkg/ui/distccl/BUILD.bazel]: Done

File: pkg/ui/distoss/BUILD.bazel
Comment (koorosh) [reply pkg/ui/distoss/BUILD.bazel]: Done

File: pkg/ui/workspaces/db-console/src/app.tsx
Comment (koorosh) [reply pkg/ui/workspaces/db-console/src/app.tsx:19]: Oh, this should be fixed before merging this PR.

File: pkg/ui/workspaces/cluster-ui/webpack.config.js
Comment (koorosh) [reply pkg/ui/workspaces/cluster-ui/webpack.config.js]: Done, I added condition in webpack configuration to exclude this change for bazel build only. Additional `is_bazel_build` env variable is passed to coordinate this flow.

**Comment by koorosh** [comment]
Updated current PR with following changes:
- webpack configuration for `db-console` and `cluster-ui` aren't changed for builds with `make`, and it's extended in a way to consume extra env variables to alter configuration required for Bazel
- fixed react router navigation for bazel build
- resolved conflicting dependency (`react-select`) between `cluster-ui` and `db-console`

**Review by rickystewart — APPROVED** [review]
Works for me, we have to take care of that "genrule outputs a directory" thing though.

**Comment by koorosh** [comment]
@rickystewart , the only change is added after your last review is version update for `build_bazel_rules_nodejs` (d366c83bfefeeb4ac6798f2407d749d7eb022792) and minor code formatting.

It is possible to build `cockroach` binary with following steps:
- start docker container with `dev builder`
- install `node_modules` as a self-managed external repository: `bazel run @nodejs//:yarn_node_repositories`
- build binary: `bazel build //pkg/cmd/cockroach` or `bazel build //pkg/cmd/cockroach-oss`

File: pkg/ui/workspaces/db-console/BUILD.bazel
Comment (rickystewart) [inline pkg/ui/workspaces/db-console/BUILD.bazel:1]: Nit: put a blank line after this line (won't pass CI unless you do)

File: pkg/ui/distccl/BUILD.bazel
Comment (rickystewart) [inline pkg/ui/distccl/BUILD.bazel:3]: Is there a bug filed for this?

File: WORKSPACE
Comment (rickystewart) [inline WORKSPACE:72]: ? Should be deleted?

**Review by rickystewart — APPROVED** [review]
Works for me, let's get this landed so we can start playing with it.

File: pkg/ui/distccl/BUILD.bazel
Comment (koorosh) [reply pkg/ui/distccl/BUILD.bazel:3]: Thanks for reminder! 
https://github.com/cockroachdb/cockroach/issues/70348

File: WORKSPACE
Comment (koorosh) [reply WORKSPACE:72]: Done

**Pull request merged**
